### ISSUE TITLE
Jerky scrolling of collectionView

### ISSUE THREAD

**Issue by JarvisTheAvenger** [krimpedance/KRPullLoader#12]
I have added KRPullLoader to the collection view. its looking jerky during scrolling.
Please let me know the solution, I am loading more data on the bottom pull.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Is there possibility of displayed cells?
What happens when nothing is displayed on cell?

**Comment by JarvisTheAvenger** [comment]
There will be the possibility of the displayed cell, I am not getting why it's jerking during scrolling.
I have also added your delegate implementations in the main queue to see if there is another thread running but I am not getting whats wrong with it.

**Comment by JarvisTheAvenger** [comment]
@krimpedance I am fetching data from server to show it on collection view

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Probably the library will not matter...
Are you getting cell data synchronously? (i.e. `UIImage(data: Data(contentsOf: URL))`)

**Comment by JarvisTheAvenger** [comment]
No, its async operation.

      DispatchQueue.main.async(execute: {
                                self.image = UIImage(data: data)
                            })

**Comment by JarvisTheAvenger** [comment]
Here's what I did in the code 

Step1- 

In viewDidLoad() 

     let loadMoreView = KRPullLoadView()
     loadMoreView.delegate = self
     cvCards.addPullLoadableView(loadMoreView, type: .loadMore)

Step 2:

Implemented KRPullloader delegates:

    extension GPCardSelectionVC: KRPullLoadViewDelegate {
    func pullLoadView(_ pullLoadView: KRPullLoadView,
                      didChangeState state: KRPullLoaderState,
                      viewType type: KRPullLoaderType) {
        if type == .loadMore {
            switch state {
            case let .loading(completionHandler):

                startIndex = arrDatasource.count
                loadMoreData(fromIndex: startIndex, limit: limit, completionHandler: completionHandler)
                
            default: break
            }
            return
        }
        
        switch state {
        case .none:
            pullLoadView.messageLabel.text = ""
            
        case let .pulling(offset, threshould):
            if offset.y > threshould {
                pullLoadView.messageLabel.text = "Pull more"
            } else {
                pullLoadView.messageLabel.text = "Release to refresh"
            }
            
        case let .loading(completionHandler):
            completionHandler()
        }
    }
}


Step 3-

In loadMoreData I have API call with parameters


Step 4:

After success of API call I am downloading images asynchronously

 DispatchQueue.main.async(execute: {
                            self.image = UIImage(data: data)
                        })


Please let me know whats wrong with the code, I tried commenting library code and run an app its working fine but when I try to run it with library its jerky during scrolling.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
ummm...
Please also show me `collectionView:cellForItemAtIndexPath:` method.

**Comment by JarvisTheAvenger** [comment]
func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let cell = collectionView .dequeueReusableCell(withReuseIdentifier: "GPCardCell", for: indexPath) as? GPCardCell else {
            return UICollectionViewCell()
        }
        
        let card = arrDatasource[indexPath.row]
        cell.configureCell(withCard: card)
        return cell
    }

**Comment by JarvisTheAvenger** [comment]
func configureCell(withCard card: Card) {
        self.card = card
        
        lblName.text = card.name.capitalized
        lblPrice.text = "$\(card.price.stringWith2Decimals())"
        
        downloadCard(card,
                     pageIndex: 0)
    }

**Comment by JarvisTheAvenger** [comment]
private func downloadCard(_ card: Card, pageIndex index: Int) {
        //download card image
        let strURL = GPManager.generateCardImageAssetURL(fromCard: card,
                                                         forPageIndex: index)
        
        imgVCard.imageFromURL(string: strURL,
                              placeholderImageName: "placeholder",
                              completion: nil)
    }

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Please try after `downloadCard` commenting out.
If it's working fine, cause is this.
Please call `downloadCard` with other thread (not main thread).

**Comment by JarvisTheAvenger** [comment]
Ok, I will try and let you know.

**Comment by JarvisTheAvenger** [comment]
@krimpedance  I tried commenting downloadCard but it's not working. I don't think that is the problem.
Is there any heavy thing you are doing in the main thread?

**Comment by JarvisTheAvenger** [comment]
@krimpedance  This is what I am doing in downloadCard method:

       public func imageFromURL(string: String?, placeholderImageName: String?, completion: ((Bool) -> Void)?) {
        if let placeholderImageName = placeholderImageName {
            self.image = UIImage(named: placeholderImageName)
        }
        else {
            self.image = nil
        }
        
        //Set unique identifier to check re-usability condition
        let strUniqueIdentifier = "\(abs(string?.hashValue ?? (UUID().uuidString.hashValue)))"
        if self.accessibilityLabel != strUniqueIdentifier {
            if let request = self.accessibilityElements?.first as? DataRequest {
                //A download task is already in progress so cancel it
                request.cancel()
                self.accessibilityElements?.removeAll()
            }
        }
        self.accessibilityLabel = strUniqueIdentifier
        
        guard let url = URL(string: string?.replacingOccurrences(of: " ", with: "%20") ?? "")
            else {
                print("\(#function)No url specified")
                
                if let completion = completion {
                    completion(false)
                }
                return
        }
        
        let activityView = addActivityIndicator(toView: self)
        var mutableRequest = URLRequest(url: url)
        mutableRequest.cachePolicy = NSURLRequest.CachePolicy.returnCacheDataElseLoad
        let request =
            Alamofire
                .request(mutableRequest)
                .response { (response) in
                    DispatchQueue.main.async(execute: {
                        activityView.removeFromSuperview()
                    })
                    
                    var success = false
                    
                    if response.response?.statusCode == 200 {
                        
                        let strUniqueIdentifier_Closure = self.accessibilityLabel
                        if strUniqueIdentifier == strUniqueIdentifier_Closure {
                            success = true
                        }
                        else {
                            success = false
                            debugPrint("Unable to set image: uniqueIdentifier check failed")
                        }
                        
                        if success, let data = response.data {
                            DispatchQueue.main.async(execute: {
                                self.image = UIImage(data: data)
                            })
                        }
                        
                    }
                    
                    if let completion = completion {
                        completion(success)
                    }
        }
        self.accessibilityElements = [request]
    }

**Comment by krimpedance** [comment]
@JarvisTheAvenger 

> Is there any heavy thing you are doing in the main thread?
Maybe no.

I will try, too.
Please show the movie if possible.

**Comment by JarvisTheAvenger** [comment]
Just try to load downloaded images async with KRPullLoader in the bottom.

**Comment by JarvisTheAvenger** [comment]
@krimpedance  The issue is in your code when I tried to remove your delegate confirmation The jerk is gone. I can also see jerk on your demo project. Please check your code once and let me know.

**Comment by JarvisTheAvenger** [comment]
@krimpedance  

      func pullLoadView(_ pullLoadView: KRPullLoadView,
                      didChangeState state: KRPullLoaderState,
                      viewType type: KRPullLoaderType) {

      }

This method is getting called every time when I scroll collection view maybe this is the problem.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Thanks for your suggestion.
I will fix it as soon as it can be confirmed.

**Comment by JarvisTheAvenger** [comment]
Any updates on issue @krimpedance

**Comment by krimpedance** [comment]
Sorry.
I'm busy recently, so can't correspond yet.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Please try [dev](https://github.com/krimpedance/KRPullLoader/tree/dev) branch.

**Comment by JarvisTheAvenger** [comment]
sure I will let you know.

**Comment by JarvisTheAvenger** [comment]
@krimpedance I think the issue is fixed now. There are no more jerky scrolling. Thanks for the help.

**Comment by JarvisTheAvenger** [comment]
@krimpedance  Can I update the pod now??

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Use this:
```
pod 'KRPullLoader', :git => 'https://github.com/krimpedance/KRPullLoader.git', :branch => 'dev'
```

**Comment by JarvisTheAvenger** [comment]
@krimpedance  when you are going to merge it with master??

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
When correction has been completed, I'd merge.
So please check it once.

**Comment by JarvisTheAvenger** [comment]
sure

**Comment by krimpedance** [comment]
@JarvisTheAvenger Any updates?

**Comment by JarvisTheAvenger** [comment]
@krimpedance  Working fine.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Thanks ;)
All right, I will merge.

**Comment by JarvisTheAvenger** [comment]
@krimpedance  please do let me know after a successful merge.

**Comment by krimpedance** [comment]
@JarvisTheAvenger 
Now available on `1.1.2` :)
Thank you.
